File: scripts/verify-decode.py
#!/usr/bin/env python3
"""
Step 3 of the build order: prove src/worker/yunet.ts decodes YuNet correctly.

A wrong decode does not crash -- it produces plausible-looking boxes that are
slightly off, and you lose a day blaming the overlay maths. So we pin it to the
reference implementation: run cv2.FaceDetectorYN over a test image, dump the
raw ONNX tensors for the *same* input, and let scripts/verify-decode.mjs run
the real TypeScript decode over those tensors and diff the two.

    pip install onnx onnxruntime opencv-python-headless numpy
    python scripts/verify-decode.py && node scripts/verify-decode.mjs
"""
import json, pathlib, sys
import numpy as np
import cv2
import onnxruntime as ort

ROOT = pathlib.Path(__file__).resolve().parent.parent
MODEL = ROOT / "resources/models/yunet.onnx"
OUT = ROOT / "test/fixtures/decode-case.json"

CONF, NMS, TOPK = 0.6, 0.3, 500


def build_scene():
    """
    One face is not a test. Tile lena at several scales into a 512x288 canvas
    so the case exercises all three strides, multi-face NMS, and faces near the
    frame edge -- the places where an off-by-one in the anchor convention shows.
    """
    src = cv2.imread(str(ROOT / "test/fixtures/lena.jpg"))
    if src is None:
        sys.exit("missing test/fixtures/lena.jpg")
    canvas = np.full((288, 512, 3), 24, np.uint8)
    for x, y, size in [(10, 20, 220), (270, 40, 150), (400, 150, 96), (180, 170, 110)]:
        tile = cv2.resize(src, (size, size), interpolation=cv2.INTER_AREA)
        h = min(size, 288 - y)
        w = min(size, 512 - x)
        canvas[y : y + h, x : x + w] = tile[:h, :w]
    return canvas


def main():
    img = build_scene()
    h, w = img.shape[:2]
    assert h % 32 == 0 and w % 32 == 0, "input must sit on the stride grid"
    cv2.imwrite(str(ROOT / "test/fixtures/scene.png"), img)

    det = cv2.FaceDetectorYN_create(str(MODEL), "", (w, h), CONF, NMS, TOPK)
    _, faces = det.detect(img)
    faces = [] if faces is None else faces.tolist()
    print(f"OpenCV reference: {len(faces)} faces in {w}x{h}")

    # Exactly what §5.1 specifies: NCHW, BGR, [0,255], no normalisation.
    # cv2.imread already gives BGR, so this is a plain transpose.
    x = img.transpose(2, 0, 1)[None].astype(np.float32)

    sess = ort.InferenceSession(str(MODEL), providers=["CPUExecutionProvider"])
    names = [o.name for o in sess.get_outputs()]
    outs = dict(zip(names, sess.run(None, {"input": x})))

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(
        json.dumps(
            {
                "inW": w,
                "inH": h,
                "conf": CONF,
                "nms": NMS,
                # OpenCV row: [x, y, w, h, 10 keypoint coords, score]
                "reference": [
                    {
                        "x": f[0], "y": f[1], "w": f[2], "h": f[3],
                        "pts": f[4:14], "score": f[14],
                    }
                    for f in faces
                ],
                "tensors": {k: np.asarray(v).ravel().round(6).tolist() for k, v in outs.items()},
            }
        )
    )
    # A small companion file the in-app self-test asserts against, without
    # dragging 950 KB of raw tensors into the renderer.
    expected = OUT.parent / "scene-expected.json"
    expected.write_text(
        json.dumps(
            {
                "width": w,
                "height": h,
                "faces": [
                    {"x": f[0], "y": f[1], "w": f[2], "h": f[3], "score": f[14]}
                    for f in faces
                ],
            },
            indent=2,
        )
    )
    print(f"wrote {OUT.relative_to(ROOT)} ({OUT.stat().st_size // 1024} KB)")
    print(f"wrote {expected.relative_to(ROOT)}")
    for f in faces:
        print(f"  box=({f[0]:7.2f},{f[1]:7.2f},{f[2]:6.2f},{f[3]:6.2f}) score={f[14]:.3f}")


if __name__ == "__main__":
    main()

File: scripts/verify-decode.mjs
#!/usr/bin/env node
// Runs the real src/worker/yunet.ts decode over the tensors dumped by
// scripts/verify-decode.py and diffs it against OpenCV's FaceDetectorYN.
// Any drift here is a decode bug, not an overlay bug.
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { fileURLToPath, pathToFileURL } from 'node:url';
import esbuild from 'esbuild';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const CASE = path.join(ROOT, 'test/fixtures/decode-case.json');
if (!fs.existsSync(CASE)) {
  console.error('missing test/fixtures/decode-case.json -- run scripts/verify-decode.py first');
  process.exit(2);
}

// Compile the TS through esbuild (already present via vite) so we test the
// shipped source, not a transcription of it. The JS API rather than the .bin
// shim: execFileSync cannot run an extensionless shim on Windows.
const tmp = path.join(os.tmpdir(), `yunet-verify-${process.pid}.mjs`);
await esbuild.build({
  entryPoints: [path.join(ROOT, 'src/worker/yunet.ts')],
  format: 'esm',
  platform: 'neutral',
  outfile: tmp,
  logLevel: 'warning',
});
// pathToFileURL, not a `file://` template: a Windows path needs escaping.
const { decode, fitLetterbox, unletterbox } = await import(pathToFileURL(tmp).href);
fs.rmSync(tmp, { force: true });

const c = JSON.parse(fs.readFileSync(CASE, 'utf8'));
const tensors = Object.fromEntries(
  Object.entries(c.tensors).map(([k, v]) => [k, Float32Array.from(v)])
);

const mine = decode(tensors, c.inW, c.inH, c.conf, c.nms);
const ref = c.reference;

console.log(`reference ${ref.length} faces | decode() ${mine.length} faces\n`);

const iou = (a, b) => {
  const x1 = Math.max(a.x, b.x),
    y1 = Math.max(a.y, b.y);
  const x2 = Math.min(a.x + a.w, b.x + b.w),
    y2 = Math.min(a.y + a.h, b.y + b.h);
  const i = Math.max(0, x2 - x1) * Math.max(0, y2 - y1);
  return i / (a.w * a.h + b.w * b.h - i);
};

let fail = 0;
const used = new Set();
for (const r of ref) {
  let best = -1,
    bestIoU = 0;
  mine.forEach((m, i) => {
    if (used.has(i)) return;
    const v = iou(r, m);
    if (v > bestIoU) {
      bestIoU = v;
      best = i;
    }
  });
  if (best < 0 || bestIoU < 0.98) {
    console.log(`  MISS box=(${r.x.toFixed(1)},${r.y.toFixed(1)}) best IoU=${bestIoU.toFixed(4)}`);
    fail++;
    continue;
  }
  used.add(best);
  const m = mine[best];
  const kpErr = Math.max(...r.pts.map((p, i) => Math.abs(p - m.pts[i])));
  const scoreErr = Math.abs(r.score - m.score);
  const ok = kpErr < 0.02 && scoreErr < 1e-3;
  if (!ok) fail++;
  console.log(
    `  ${ok ? 'OK  ' : 'FAIL'} box=(${r.x.toFixed(1)},${r.y.toFixed(1)},${r.w.toFixed(1)},${r.h.toFixed(1)}) ` +
      `IoU=${bestIoU.toFixed(5)} maxKpErr=${kpErr.toFixed(4)}px scoreErr=${scoreErr.toExponential(1)}`
  );
}
const extra = mine.length - used.size;
if (extra > 0) {
  console.log(`  ${extra} unmatched detection(s) from decode()`);
  fail += extra;
}

// unletterbox must be the exact inverse of fitLetterbox.
const lb = fitLetterbox(1280, 720, 320);
console.log(
  `\nletterbox 1280x720 -> ${lb.inW}x${lb.inH} (fit ${lb.fitW}x${lb.fitH}, pad ${lb.padX},${lb.padY})`
);
const probe = [
  { x: lb.padX, y: lb.padY, w: lb.fitW, h: lb.fitH, score: 1, pts: new Array(10).fill(lb.padX) },
];
const back = unletterbox(probe, lb)[0];
const roundTripOk =
  Math.abs(back.x) < 0.51 &&
  Math.abs(back.y) < 0.51 &&
  Math.abs(back.w - 1280) < 2 &&
  Math.abs(back.h - 720) < 2;
console.log(
  `  full-frame box round-trips to (${back.x.toFixed(2)},${back.y.toFixed(2)},${back.w.toFixed(1)},${back.h.toFixed(1)}) ${roundTripOk ? 'OK' : 'FAIL'}`
);
if (!roundTripOk) fail++;

console.log(fail === 0 ? '\nPASS' : `\nFAIL (${fail})`);
process.exit(fail === 0 ? 0 : 1);
